test/e2e/deployment.go

```diff
@@ -271,3 +335,178 @@ func (d *Deployment) WaitForEnvoyDaemonSetUpdated() error {
+		return err
+	}
+
+	bFile, err := ioutil.TempFile("", "bootstrap-*.json")
+	if err != nil {
+		return err
+	}
+
+	// Generate bootstrap config with Contour local address and plaintext
+	// client config.
+	bootstrapCmd := exec.Command( // nolint:gosec
+		d.contourBin,
+		"bootstrap",
+		bFile.Name(),
+		"--xds-address="+d.localContourHost,
+		"--xds-port="+d.localContourPort,
+		"--xds-resource-version=v3",
+	)
+	session, err := gexec.Start(bootstrapCmd, d.cmdOutputWriter, d.cmdOutputWriter)
+	if err != nil {
+		return err
+	}
+	session.Wait()
+
+	bootstrapContents, err := ioutil.ReadAll(bFile)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		bFile.Close()
+		os.RemoveAll(bFile.Name())
+	}()
+
+	bootstrapConfigMap := &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "envoy-bootstrap",
+			Namespace: d.Namespace.Name,
+		},
+		Data: map[string]string{
+			"envoy.json": string(bootstrapContents),
+		},
+	}
+	if err := d.ensureResource(bootstrapConfigMap, new(v1.ConfigMap)); err != nil {
+		return err
+	}
+
+	// Add bootstrap ConfigMap as volume on Envoy pods (also removes cert volume).
+	d.EnvoyDaemonSet.Spec.Template.Spec.Volumes = []v1.Volume{
+		{
+			Name: "envoy-config",
+			VolumeSource: v1.VolumeSource{
+				ConfigMap: &v1.ConfigMapVolumeSource{
+					LocalObjectReference: v1.LocalObjectReference{
+						Name: "envoy-bootstrap",
+					},
+				},
+			},
+		},
+	}
+
+	// Remove cert volume mount.
+	d.EnvoyDaemonSet.Spec.Template.Spec.Containers[1].VolumeMounts = []v1.VolumeMount{d.EnvoyDaemonSet.Spec.Template.Spec.Containers[1].VolumeMounts[0]}
+
+	// Remove init container.
+	d.EnvoyDaemonSet.Spec.Template.Spec.InitContainers = nil
+	// Remove shutdown-manager container.
+	d.EnvoyDaemonSet.Spec.Template.Spec.Containers = d.EnvoyDaemonSet.Spec.Template.Spec.Containers[1:]
+
+	if err := d.EnsureEnvoyDaemonSet(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Starts local contour, applying arguments and marshaling config into config
+// file. Returns running Contour command and config file so we can clean them
+// up.
+func (d *Deployment) StartLocalContour(config *config.Parameters, additionalArgs ...string) (*gexec.Session, string, error) {
+	configFile, err := ioutil.TempFile("", "contour-config-*.yaml")
+	if err != nil {
+		return nil, "", err
+	}
+	defer configFile.Close()
+
+	content, err := yaml.Marshal(config)
+	if err != nil {
+		return nil, "", err
+	}
+	if err := os.WriteFile(configFile.Name(), content, 0644); err != nil {
+		return nil, "", err
+	}
+
+	contourServeArgs := append([]string{
+		"serve",
+		"--xds-address=0.0.0.0",
+		"--xds-port=" + d.localContourPort,
+		"--insecure",
+		"--kubeconfig=" + d.kubeConfig,
+		"--config-path=" + configFile.Name(),
+	}, additionalArgs...)
+	session, err := gexec.Start(exec.Command(d.contourBin, contourServeArgs...), d.cmdOutputWriter, d.cmdOutputWriter) // nolint:gosec
+	if err != nil {
+		return nil, "", err
+	}
+	return session, configFile.Name(), nil
```

Comment:
was thinking about making a struct to hold the things needed for a local contour session, the gexec.Session object and config file path etc.